### ISSUE TITLE
Bind mounts under volume are not cleaned or reused on container stop/restart

### ISSUE THREAD

**Issue by bengerman13** [moby/moby#36179]
**Description**
I'm using an image with volume declarations, and I'm bind mounting sub-directories of that volume on an nfs mount with rshared propagation. When I do this, mounts are created on the host, but they are not cleaned up by Docker when the container is stopped or removed, and they are duplicated each time the container is restarted.

**Steps to reproduce the issue:**
1. Bind mount a directory in a volume with shared or rshared propagation

**Describe the results you received:**
```
~/docker$ ls
Dockerfile
~/docker$ cat Dockerfile 
from bash

VOLUME /myvol
~/docker$ docker build -t bash_mount .
Sending build context to Docker daemon  2.048kB
Step 1/2 : from bash
 ---> 16c7ca5e7283
Step 2/2 : VOLUME /myvol
 ---> Using cache
 ---> 28bf51cf631d
Successfully built 28bf51cf631d
Successfully tagged bash_mount:latest
~/docker$ docker run -d --mount type=bind,source=${MOUNTPOINT},target=/myvol/mount,bind-propagation=rshared --name test bash_mount sleep 10000000 
6f4521025544bbb709d8e2ade4de448dc924558c97f0324950ee6500f82adebe
~/docker$ docker inspect test -f '{{range  .Mounts }}{{ .Name }}{{end}}'
50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12
~/docker$ mount | grep 50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12 | wc -l
1
~/docker$ docker restart test
test
~/docker$ mount | grep 50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12 | wc -l
2
~/docker$ docker restart test
test
~/docker$ mount | grep 50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12 | wc -l
4
~/docker$ docker stop test
test
~/docker$ mount | grep 50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12 | wc -l
4
~/docker$ docker rm test
test
~/docker$ mount | grep 50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12 | wc -l
4
~/docker$ docker volume prune
WARNING! This will remove all volumes not used by at least one container.
Are you sure you want to continue? [y/N] y
Total reclaimed space: 0B
~/docker$ mount | grep 50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12 | wc -l
4
~/docker$ docker volume rm 
06e9dba21363ef29b2bed01db52bf23aa39dd0c0a4687636fe3fc5e6085e2c0b  5be4aba0d3f7a7c9de9f174f193d7b997b99af13d4fb8e9852db2438e3a7e6a9  
50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12  608af439dbc4de46b86acc481d28d9e1f592d1e7ca4f226f7583055e8bbf6b90  
~/docker$ docker volume rm 50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12
Error response from daemon: unable to remove volume: remove 50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12: error removing volume path '/var/lib/docker/volumes/50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12/_data': remove /var/lib/docker/volumes/50ca458df54162f34aea0079431e58dab9a557f339be0f2ba6de1f85b9683e12/_data/mount: device or resource busy
```

**Describe the results you expected:**
The bind mounts should not duplicate, and should be cleaned up by removing the volume or container. 

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `docker version`:**

```
~/docker$ docker version
Client:
 Version:	17.12.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	c97c6d6
 Built:	Wed Dec 27 20:11:19 2017
 OS/Arch:	linux/amd64

Server:
 Engine:
  Version:	17.12.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	c97c6d6
  Built:	Wed Dec 27 20:09:53 2017
  OS/Arch:	linux/amd64
  Experimental:	false
```

**Output of `docker info`:**

```
~/docker$ docker info
Containers: 8
 Running: 8
 Paused: 0
 Stopped: 0
Images: 26
Server Version: 17.12.0-ce
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 148
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 89623f28b87a6004d4b785663257362d1658a729
runc version: b2567b37d7b75eb4cf325b77297b140ea686ce8f
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-112-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 12
Total Memory: 39.19GiB
Name: dev-aics-bbp-001
ID: XT2J:Q26X:D6EA:M7RQ:XLRB:FAAO:5WSG:UCOX:46ZT:2ON2:L3UN:W47V
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Physical

**Comment by cpuguy83** [comment]
This is working as expected.
The mount is being propagated to the host namespace. If you don't want it to do this, then don't use shared propagation.
It is up to the containerized process to clean this up properly.

**Comment by bengerman13** [comment]
@cpuguy83 I'm not convinced this is a moby/docker bug necessarily, but your comment doesn't make sense to me. 
How is the containerized process supposed to clean up the mounts? Isn't docker creating the bind mount? The duplicated mounts are all bind mounts from the host to `/var/lib/docker/volumes/$VOLUME_ID/_data/mount`

**Comment by cpuguy83** [comment]
So this is creating a volume on `/myvol`, then you are creating a bindmount from the host to `/myvol/mount`.
The `/mount` bit is propagating to the host mount namespace due to the use of `rshared`.

Docker could make sure everything is unmounted in the volume path before attempting to remove for sure... but also this is basically a situation of "don't do that".

I'll leave this open so that we can track implementing that unmount.

**Comment by kvaps** [comment]
Agree, same problem with
```
# docker version
Client:
 Version:      17.03.2-ce
 API version:  1.27
 Go version:   go1.7.5
 Git commit:   f5ec1e2
 Built:        Tue Jun 27 03:35:14 2017
 OS/Arch:      linux/amd64

Server:
 Version:      17.03.2-ce
 API version:  1.27 (minimum version 1.12)
 Go version:   go1.7.5
 Git commit:   f5ec1e2
 Built:        Tue Jun 27 03:35:14 2017
 OS/Arch:      linux/amd64
 Experimental: false
```

Simple steps for reproduce:
```shell
# docker run -ti --rm -v /tmp/mounts:/mounts:rshared -v /tmp/test1:/mounts/test1 alpine true
# mountpoint /tmp/mounts/test1
/tmp/mounts/test1 is a mountpoint
# docker run -ti --rm -v /tmp/mounts:/mounts:rshared -v /tmp/test1:/mounts/test1 alpine true
# mountpoint /tmp/mounts/test1
/tmp/mounts/test1 is a mountpoint
# umount /tmp/mounts/test1
# mountpoint /tmp/mounts/test1
/tmp/mounts/test1 is a mountpoint
# umount /tmp/mounts/test1
# mountpoint /tmp/mounts/test1
/tmp/mounts/test1 is not a mountpoint
```
```

**Comment by jrray** [comment]
Apologies for adding to this oldish ticket, but we encountered a data loss situation related to the mount scenario described in here.

With the live mount left behind in `/var/lib/docker/volumes/$VOLUME_ID/_data/mount`, and the container killed/deleted, `docker volume rm $VOLUME_ID` will traverse the `mount` mountpoint and delete everything there. This was unexpected behavior. Ideally `volume rm` would have "--one-file-system" behavior to prevent this.

We realize that using "slave" instead of "shared" bind propagation will avoid this issue.

What added insult to injury is that we were using the `ro` option on the shared mount, but even with `ro` any new mounts appearing under this path are still mounted with `rw`.

Tested with docker-ce-18.03.1.ce-1 on CentOS 7.5.1804.
